x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/action/cohere/CohereEmbeddingsAction.java

```diff
@@ -38,13 +39,13 @@ public CohereEmbeddingsAction(Sender sender, CohereEmbeddingsModel model) {
     }
 
     @Override
-    public void execute(List<String> input, ActionListener<InferenceServiceResults> listener) {
+    public void execute(@Nullable String query, List<String> input, ActionListener<InferenceServiceResults> listener) {
```

Comment:
This is a reminder for my future self haha. If you don't have time to do it in this PR that's fine, we can refactor it later. I think what we could do here is leverage the visitor pattern and create an interface like `Input` that accepts a visitor that is a "request creator" for a particular external service which has visit methods for each type of input we'd get like, `RerankerInput` `EmbeddingsInput` etc. The request creator would already be for cohere for example and then know it needs to pass `RerankerInput` on to the specific reranker request creator.

That way instead of needing to include `query` and `input` here we'd just include an interface `Input` and hopefully we wouldn't have to change this methods again. We'd just add more methods to the `Input` class and create the corresponding visitors.

I'm open to other ideas too though 👍

Reply:
That sounds like a smart solution, but I need to think about it more. I'll try to include this change in my PR.